### PR TITLE
628 make text to centreline case insensitive

### LINKED ISSUE #628: Make Text to Centreline Case Insensitive
From https://github.com/CityofToronto/bdit_data-sources/pull/421

> Currently, gis.text_to_centreline() fails if the input text is capitalized, e.g.:
> 
> ```sql
> SELECT *
> FROM gis.text_to_centreline(111,'BLOOR ST W', 'BETWEEN SHAW ST AND OSSINGTON AVE', NULL);
> ```
> The problem has to do with using SPLIT_PART on fixed lower-case text and in the function function-clean_bylaws_text.sql which gis.text_to_centreline() calls. The 'Between' is also only checked for first letter capitalization.

Unfortunately that PR introduced more complexity to an already rather complicated set of functions. Opening this issue for further improvements.

### PR DESCRIPTION
**Pull request by gabrielwol**
## What this pull request accomplishes:

3 modified and 1 new function to make `text_to_centreline` case insensitive.
- new function: `custom_case`: new function to selectively capitalize/lowercase certain words
- `gis._clean_bylaws_text`: lots of whitespace changes, call `custom_case` on input params, add some variables to simplify a few blocks. 
- `abbr_street`: make some replacements case insensitive
- `_translate_intersection_point`: match lower or upper and abbreviated or full cardinal directions

Also added a bash script for testing: `gis/text_to_centreline/sql/test_script.sh`
- recreates everything in a new schema
- then renames back to `gis` for committing!

## Issue(s) this solves:

- Closes #628 

## What, in particular, needs to reviewed:

- Lots of changes but see test cases below! Seems to be working!

## What needs to be done by a sysadmin after this PR is merged

_E.g.: these tables need to be migrated/created in the production schema._

### CODE REVIEW

**Comment by gabrielwol** [comment]
The four test cases from the readme are now working in both upper and lower case! 

```sql
SELECT tests.*, clean.*
FROM (
    VALUES
        (1, 'RICHLAND CRESCENT', 'ENTIRE LENGTH', NULL),
        (2, 'FOXRIDGE DRIVE', 'BIRCHMOUNT ROAD AND KENNEDY ROAD', NULL),
        (3, 'RAVENWOOD PLACE', 'FERRIS ROAD AND A POINT 64.11 METRES SOUTHWEST', NULL),
        (4, 'BRIMORTON DRIVE', 'A POINT 101 METRES WEST OF AMBERJACK BOULEVARD AND A POINT 103 METRES EAST OF DOLLY VARDEN BOULEVARD', NULL)
        --(1, 'Richland Crescent', 'Entire Length', NULL),
        --(2, 'Foxridge Drive', 'Birchmount Road and Kennedy Road', NULL),
        --(3, 'Ravenwood Place', 'Ferris Road and a point 64.11 metres southwest', NULL),
        --(4, 'Brimorton Drive', 'A point 101 metres west of Amberjack Boulevard and a point 103 metres east of Dolly Varden Boulevard', NULL)
    ) AS tests (_bylaw_id, highway, frm, "to"),
LATERAL (
    SELECT *
    FROM gwolofs.text_to_centreline(tests._bylaw_id, tests.highway, tests.frm, tests."to")
) AS clean
ORDER BY 2, 1
```

File: gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql [hunk 1]
```diff
@@ -4,17 +4,27 @@ CREATE OR REPLACE FUNCTION gis._translate_intersection_point(
...
+    CASE
+        WHEN regexp_instr(direction, '(?<!\S)(w|west)(?!\S)') = 1
+        THEN ST_Translate(oid_geom, -metres, 0)
+        WHEN regexp_instr(direction, '(?<!\S)(e|east)(?!\S)') = 1
+        THEN ST_Translate(oid_geom, metres, 0)
+        WHEN regexp_instr(direction, '(?<!\S)(n|north)(?!\S)') = 1
+        THEN ST_Translate(oid_geom, 0, metres)
+        WHEN regexp_instr(direction, '(?<!\S)(s|south)(?!\S)') = 1
+        THEN ST_Translate(oid_geom, 0, -metres)
+        WHEN direction = 'southwest'
```
Comment (chmnata) [inline gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql:18]: is it ok to get rid of possibility of matching to ones with space in between words? e.g south west

File: gis/text_to_centreline/sql/helper_functions/function-abbr_street.sql [hunk 2]
```diff
@@ -41,11 +41,11 @@ _abbrev_street := regexp_REPLACE(_abbrev_street, '(?<=[A-Z][a-z]+ )(Park)(?![a-z
...
-_abbrev_street := regexp_REPLACE(_abbrev_street, 'approximately', '', 'g');
+_abbrev_street := regexp_REPLACE(_abbrev_street, 'approximately', '', 'gi');
 _abbrev_street := regexp_REPLACE(_abbrev_street, ' Grove', ' Grv', 'g');
 _abbrev_street := regexp_REPLACE(_abbrev_street, ' Grv Rd', ' Grove Rd', 'g');
 _abbrev_street := regexp_REPLACE(_abbrev_street, ' Grv Ave', ' Grove Ave', 'g');
 _abbrev_street := regexp_REPLACE(_abbrev_street, 'Rd Cres', 'Road Cres', 'g');
 return _abbrev_street;
 end;
-$BODY$;
+$BODY$;
```
Comment (chmnata) [inline gis/text_to_centreline/sql/helper_functions/function-abbr_street.sql:51]: there is one permission difference between this and the old function, not sure what this is but we should add it as well
```
GRANT EXECUTE ON FUNCTION gis.abbr_street_stop_using_this(text) TO vz_readonly;
```

File: gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql (hunk 1 above)
Comment (chmnata) [reply gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql:18]: the translated point is different, also some weird results from the one with space between north and east :meow_dio:
```
SELECT tests.*, clean.*
FROM (
    VALUES
		(4, 'BRIMORTON DRIVE', 'A POINT 101 METRES south WEST OF AMBERJACK BOULEVARD AND A POINT 103 METRES north East OF DOLLY VARDEN BOULEVARD', NULL),
        (5, 'BRIMORTON DRIVE', 'A POINT 101 M southwest OF AMBERJACK BOULEVARD AND A POINT 103 M northeast OF DOLLY VARDEN BOULEVARD', NULL)
        --(1, 'Richland Crescent', 'Entire Length', NULL),
        --(2, 'Foxridge Drive', 'Birchmount Road and Kennedy Road', NULL),
        --(3, 'Ravenwood Place', 'Ferris Road and a point 64.11 metres southwest', NULL),
        --(4, 'Brimorton Drive', 'A point 101 metres soutwest of Amberjack Boulevard and a point 103 metres east of Dolly Varden Boulevard', NULL)
    ) AS tests (_bylaw_id, highway, frm, "to"),
LATERAL (
    SELECT *
    FROM gwolofs.text_to_centreline(tests._bylaw_id, tests.highway, tests.frm, tests."to")
) AS clean
ORDER BY 2, 1
```

**Review by chmnata — CHANGES_REQUESTED** [review]
Thanks for tackling this!! Glad to see text to centreline can now be applied to many more cases. Mainly just one fix to add space between south and west to enable that matching. The rest of semi-confusing excluding of directions we talked about is not the point of this PR so we can merge once that is fixed

File: gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql (hunk 1 above)
Comment (gabrielwol) [reply gis/text_to_centreline/sql/helper_functions/function-translate_intersection_point.sql:18]: I adjusted the `custom_case` function to clean up all the directions with spaces in them!

Now these both work: 🥳 
```sql
WITH val (_bylaw_id, highway, frm, "to") AS (
    (VALUES 
        (4, 'BRIMORTON DRIVE', 'A POINT 101 METRES south WEST OF AMBERJACK BOULEVARD AND A POINT 103 METRES north East OF DOLLY VARDEN BOULEVARD', NULL),
        (5, 'BRIMORTON DRIVE', 'A POINT 101 M southwest OF AMBERJACK BOULEVARD AND A POINT 103 M northeast OF DOLLY VARDEN BOULEVARD', NULL)
    )
)

SELECT 'gwolofs', * FROM val, gwolofs.text_to_centreline(val._bylaw_id, val.highway, val.frm, val."to")
```

**Review by chmnata — APPROVED** [review]
thanks for the fixes! @gabrielwol

**Pull request merged**